parser_library/src/workspaces/file_manager_impl.cpp

```diff
@@ -14,56 +14,329 @@
+    {
+        i += 1 + text.substr(i).starts_with(nl);
+        output.push_back(i);
+    }
+}
 
-    std::optional<std::string> result(file->get_text());
-    if (file->is_bad())
-        result.reset();
+std::vector<size_t> create_line_indices(std::string_view text)
+{
+    std::vector<size_t> ret;
+    create_line_indices(ret, text);
+    return ret;
+}
 
-    return result;
+void create_line_indices(std::vector<size_t>& output, std::string_view text)
+{
+    output.clear();
+    output.push_back(0);
+    find_newlines(output, text);
 }
 
-void file_manager_impl::remove_file(const file_location& file)
+void apply_text_diff(std::string& text, std::vector<size_t>& lines, range r, std::string_view replacement)
 {
-    std::lock_guard guard(files_mutex);
-    if (!files_.contains(file))
+    if (r.start > r.end || r.end.line > lines.size())
         return;
 
-    // close the file internally
-    files_.erase(file);
+    size_t range_end_line = (size_t)r.end.line;
+    size_t range_start_line = (size_t)r.start.line;
+
+    size_t begin = index_from_position(text, lines, r.start);
+    size_t end = index_from_position(text, lines, r.end);
+
+    text.replace(begin, end - begin, replacement);
+
+    std::vector<size_t> new_lines;
+    find_newlines(new_lines, replacement);
+
+    size_t old_lines_count = range_end_line - range_start_line;
+    size_t new_lines_count = new_lines.size();
+
+    size_t char_diff = replacement.size() - (end - begin);
+
+    // add or remove lines depending on the difference
+    if (new_lines_count > old_lines_count)
+    {
+        size_t diff = new_lines_count - old_lines_count;
+        lines.insert(lines.end(), diff, 0);
+
+        for (size_t i = lines.size() - 1; i > range_end_line + diff; --i)
+        {
+            lines[i] = lines[i - diff] + char_diff;
+        }
+    }
+    else
+    {
+        size_t diff = old_lines_count - new_lines_count;
+
+        for (size_t i = range_start_line + 1 + new_lines_count; i < lines.size() - diff; ++i)
+        {
+            lines[i] = lines[i + diff] + char_diff;
+        }
+
+        lines.erase(lines.end() - diff, lines.end());
+    }
+
+
+    for (size_t i = range_start_line + 1; i <= range_start_line + new_lines_count; ++i)
+    {
+        lines[i] = new_lines[i - (size_t)range_start_line - 1] + begin;
+    }
+}
+
+
+struct file_manager_impl::mapped_file final : file
+{
+    // there is no noexcept version of std::enable_shared_from_this::shared_from_this
+    std::weak_ptr<mapped_file> m_self;
+    std::shared_ptr<mapped_file> shared_from_this() const noexcept { return m_self.lock(); }
+
+    file_location m_location;
+    std::string m_text;
+    struct file_error
+    {};
+    std::optional<file_error> m_error;
+    // Array of "pointers" to m_text where lines start.
+    std::vector<size_t> m_lines;
+
+    file_manager_impl& m_fm;
+
+    version_t m_lsp_version = 0;
```

Comment:
Is `m_lsp_version` actively used somewhere?
Or is it considered for some kind of bookkeeping?

Reply:
At the moment it is not used. It is something provided by the VSCode in the LSP messages, but afaik the numbers are not unique, so for now, we just keep them around without actively using them.